# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/printing/nsPrintJob.cpp
# Commit: 25fa445f9702
# Full Hash: 25fa445f9702185490edb0b13e4d950c8be2af1b
# Author: Jonathan Watt <jwatt@jwatt.org>
# Date: 2020-07-11 21:52:53
# Regressor Bug: 1557983
# File Overlap Count: 1
# Description:
#   Bug 1557983. Significantly simplify and fix the 'beforeprint'/'afterprint' dispatching code.
#   
#   This fixes two issue.
#   
#   First, the code shouldn't be dispatching these events every time it gets a new
# ==============================================================================

diff -r d4d6783f41b2 -r 25fa445f9702 layout/printing/nsPrintJob.cpp
--- a/layout/printing/nsPrintJob.cpp	Fri Jul 10 23:08:54 2020 +0000
+++ b/layout/printing/nsPrintJob.cpp	Sat Jul 11 14:14:03 2020 +0000
@@ -168,6 +168,26 @@
 #  define DUMP_DOC_TREELAYOUT
 #endif
 
+static CallState CollectDocuments(Document& aDoc,
+                                  nsTArray<nsCOMPtr<Document>>& aDocs) {
+  aDocs.AppendElement(&aDoc);
+  auto recurse = [&aDocs](Document& aSubDoc) {
+    return CollectDocuments(aSubDoc, aDocs);
+  };
+  aDoc.EnumerateSubDocuments(recurse);
+  return CallState::Continue;
+}
+
+static void DispatchEventToWindowTree(Document& aDoc, const nsAString& aEvent) {
+  nsTArray<nsCOMPtr<Document>> targets;
+  CollectDocuments(aDoc, targets);
+  for (nsCOMPtr<Document>& doc : targets) {
+    nsContentUtils::DispatchTrustedEvent(doc, doc->GetWindow(), aEvent,
+                                         CanBubble::eNo, Cancelable::eNo,
+                                         nullptr);
+  }
+}
+
 class nsScriptSuppressor {
  public:
   explicit nsScriptSuppressor(nsPrintJob* aPrintJob)
@@ -663,6 +683,14 @@
   nsCOMPtr<nsIDocShell> docShell(do_QueryReferent(mDocShell, &rv));
   NS_ENSURE_SUCCESS(rv, rv);
 
+  if (!aSourceDoc->IsStaticDocument()) {
+    // This is the original document.  We must send 'beforeprint' and
+    // 'afterprint' events to give the document the chance to make changes
+    // for print output.  (Obviously if `aSourceDoc` is a clone, it already
+    // has these changes.)
+    DispatchEventToWindowTree(*aSourceDoc, u"beforeprint"_ns);
+  }
+
   {
     nsAutoScriptBlocker scriptBlocker;
     printData->mPrintObject = MakeUnique<nsPrintObject>();
@@ -680,6 +708,10 @@
                             printData->mPrintObject, &printData->mPrintDocList);
   }
 
+  if (!aSourceDoc->IsStaticDocument()) {
+    DispatchEventToWindowTree(*aSourceDoc, u"afterprint"_ns);
+  }
+
   // The nsAutoScriptBlocker above will now have been destroyed, which may
   // cause our print/print-preview operation to finish. In this case, we
   // should immediately return an error code so that the root caller knows
